### ISSUE TITLE
React typings allow false for class components, but not for stateless

### ISSUE THREAD

**Issue by magnushiie** [DefinitelyTyped/DefinitelyTyped#18912]
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @gaspard @vbfox 

As mentioned in [React documentation](https://facebook.github.io/react/docs/conditional-rendering.html#inline-if-with-logical--operator), a stateless component can return `false` for empty content. Currently [StatelessComponent](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L312-L313) is defined as:
```ts
interface StatelessComponent<P = {}> {
        (props: P & { children?: ReactNode }, context?: any): ReactElement<any> | null;
```
whereas [Component](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L270-L280) has
```ts
class Component<P, S> {
        constructor(props?: P, context?: any);
        // ...snip...
        render(): JSX.Element | null | false;
```

Shouldn't these types be the same (including eliminating the ReactElement vs JSX.Element difference)? And to avoid future such issues (#14064, this one) it should probably be a named type, e.g. RenderResult.

A search for `| null` also comes up with `ComponentSpec.render` which probably should also have the same result type.

**Comment by magnushiie** [comment]
Note: React 16 [will allow](https://twitter.com/dan_abramov/status/805037210613534720?lang=en) arrays and strings and arrays of strings also. So at one point it would be similar to `ReactNode`.

Also, if unifying with `ReactNode`, the recursive definition can be done as described by @ahejlsberg in https://github.com/Microsoft/TypeScript/issues/3496#issuecomment-128553540

**Comment by vbfox** [comment]
I mentioned it here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17021#issuecomment-307027890 but forget to follow or PR.

**Comment by gaspard** [comment]
@vbox Using a type for this is a good idea. Please also add a test for any new elements in this return type (like `false`). :-)

**Comment by adrienharnay** [comment]
Hey, is anyone working on this? If not I could help finish this if needed :)

**Comment by mgol** [comment]
Currently, the class `Component` type specifies its `render` method type as:
```ts
render(): ReactNode;
```

Shouldn't `FunctionComponent` type be changed from:
```ts
(props: P & { children?: ReactNode }, context?: any): ReactElement<any> | null;
```
to:
```ts
(props: P & { children?: ReactNode }, context?: any): ReactNode;
```
as well? As far as I understand both return types should be identical. Current types make it hard to refactor a class component as a function one as then suddently the return types change.

**Comment by Hotell** [comment]
@Kovensky did this PR AFAIR , any reasons why FC returns `ReactElement | null` instead of `ReactNode` 

> I guess it might be related that ReactNode contains `{}` and `undefined` which makes compoennt returns type checking hard https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L171

**Comment by Kovensky** [comment]
That is because the compiler is still dumb in this respect and thinks function components **have** to return `JSX.Element | null`.

```tsx
function Test() { return null }
function Test2() { return false }

const jsx = (
  <>
    <Test />
    <Test2
      // type error
      // "JSX element type 'boolean' is not a constructor function for JSX elements."
      // (???)
    />
  </>
)
```

This is unrelated to the type of `FunctionComponent`, and correcting the type in `FunctionComponent` would actually make it impossible to use things typed or union-ed with `FunctionComponent` as JSX tags.

![image](https://user-images.githubusercontent.com/73085/49809745-075f9000-fda3-11e8-857c-084a18e1ed1e.png)

**Comment by eps1lon** [comment]
This can be fixed once Microsoft/TypeScript#21699 is resolved.

**Comment by TomasHubelbauer** [comment]
This seems to be the reason for why `React.FunctionComponent` (and `React.SFC`) cannot return `ReactNode` (because it contains `false` and `null` and `undefined`), but could they return `ReactChild` instead of `ReactElement`?

Right now this is an error:

```tsx
const test: React.FunctionComponent<{}> = props => 'test';
```

I would like to proposed that `FunctionComponent` returns `ReachChild` instead, but I am not sure whether that's the way to go. Can anyone advise?

**Comment by AviVahl** [comment]
Almost opened another duplicate... `() => 'Hello World'` should be a valid `FunctionComponent`. :/

**Comment by rodoabad** [comment]
What can we do to move forward with this?

**Comment by rodoabad** [comment]
It's 2021 and this is still an open issue, is there anything we can do to help move this forward?

**Comment by eps1lon** [comment]
> This can be fixed once [microsoft/TypeScript#21699](https://github.com/microsoft/TypeScript/issues/21699) is resolved.

This is still the latest state of this issue. We can't do anything unless TypeScript folks move forward with microsoft/TypeScript#21699

**Comment by eps1lon** [comment]
Duplicate of https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18051

**Comment by ryami333** [comment]
Can anyone explain to my why this is necessarily blocked by the mentioned Typescript issue when this is replicable without using JSX? This, for example, creates a type-error, because `FunctionComponent` is not permissive enough to allow functional components to return numbers, booleans or `undefined`.

```ts
const ComponentWhichReturnsOne = () => 1;

React.createElement(ComponentWhichReturnsOne);
```

**Comment by eps1lon** [comment]
@ryami333  By allowing it in `createElement` but not JSX we would cause even more confusion. Especially since JSX is the common way to create elements as opposed to manually calling JSX factories.

**Comment by ryami333** [comment]
Right, so it's not allowed in JSX because of technical restraints, but it's only not allowed in vanilla _because equality_?

**Comment by eps1lon** [comment]
This was just the first argument that come to mind. We already have inconsistencies between JSX and no-JSX so it may not be as relevant.

Feel free to go through the typings for `createElement` to see if a fix makes sense. We would probably end up with another overload that accepts something like "ActualFunctionComponent" that allows returning everything that's allowed at runtime. In other words, we would have to add types that we really shouldn't need but now have to maintain for an edge case. That doesn't sound like the cost-benefit ratio is great.

**Comment by bottd** [comment]
Commenting to point out that I've been running into this with my team somewhat often. It'd be great to get the type corrected at some point, the common use case of returning `false` for me is allowing returns of short circuited components.
```js
return condition && <div>My markup</div>
```
The current types force this code to be rewritten as a ternary to return `null` instead of `false`. I haven't contributed to DefinitelyTyped before, but I may try taking a look into this. It doesn't seem too promising considering this issue has been open for quite a while though.

**Comment by eps1lon** [comment]
>  It'd be great to get the type corrected at some point

This is still blocked by https://github.com/microsoft/TypeScript/issues/21699. I recommend subscribing to https://github.com/microsoft/TypeScript/issues/21699 instead of asking for updates.

**Comment by eps1lon** [comment]
`JSX.Element` is used by TypeScript itself. You essentially told TypeScript that `<div />` is a `React.ReactNode`.

**Comment by user72356** [comment]
Now that https://github.com/microsoft/TypeScript/issues/21699 is fixed, should this issue and https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60242 be fixed as well?

**Comment by eps1lon** [comment]
Still need to wait for a stable release but I prepare support for adding pre 5.1. and 5.1 support in `@types/react` and then I'm hopeful we can close since once 5.1 is out ([which is scheduled for June 2023](https://github.com/microsoft/TypeScript/issues/53031))
